File: src/routes/performance.py
"""性能追踪路由 - 接收前端性能数据上报并支持链路追踪"""
import json
import logging
import uuid
from datetime import datetime, timedelta
from flask import Blueprint, request, jsonify

from src.models.performance_trace import PerformanceTrace, PerformanceStat

logger = logging.getLogger(__name__)

# 创建蓝图
performance_bp = Blueprint('performance', __name__, url_prefix='/api/performance')


def generate_span_id():
    """生成span ID"""
    return uuid.uuid4().hex[:16]


def get_trace_type(name):
    """根据名称判断trace类型"""
    if name.startswith('web-vital_'):
        return 'web_vital'
    elif name.startswith('api_'):
        return 'api_call'
    elif name.startswith('custom_'):
        return 'function'
    elif name.startswith('component_'):
        return 'component'
    elif name.startswith('page_'):
        return 'page_load'
    else:
        return 'operation'


@performance_bp.route('', methods=['POST'])
def record_performance():
    """接收前端性能数据 - 支持链路追踪"""
    try:
        data = request.get_json()
        if not data:
            return jsonify({'success': False, 'error': 'No data provided'}), 400

        # 提取数据
        name = data.get('name', 'unknown')
        value = data.get('value', 0)
        rating = data.get('rating', 'unknown')
        page = data.get('page', 'unknown')
        timestamp = data.get('timestamp', 0)
        trace_id = data.get('trace_id')
        parent_span_id = data.get('parent_span_id')

        # 生成 span_id
        span_id = generate_span_id()

        # 确定 trace_type
        trace_type = get_trace_type(name)

        # 计算时间戳
        start_time = timestamp / 1000  # 转换为秒
        end_time = start_time + (value / 1000)

        # 保存到数据库
        trace = PerformanceTrace.create(
            trace_id=trace_id or span_id,
            parent_span_id=parent_span_id,
            span_id=span_id,
            trace_type=trace_type,
            name=name,
            page=page,
            duration=float(value),
            start_time=start_time,
            end_time=end_time,
            rating=rating,
            metadata=json.dumps(data.get('metadata', {}))
        )

        logger.info(f"[PERF] {name}={value}ms rating={rating} page={page} span_id={span_id[:8]}")

        return jsonify({
            'success': True,
            'span_id': span_id,
            'trace_id': trace.trace_id
        }), 200

    except Exception as e:
        logger.error(f"[PERF] Error recording performance: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@performance_bp.route('/batch', methods=['POST'])
def batch_record_performance():
    """批量接收性能数据 - 用于一次上报多个span"""
    try:
        data = request.get_json()
        if not data or 'spans' not in data:
            return jsonify({'success': False, 'error': 'No spans provided'}), 400

        spans = data['spans']
        trace_id = data.get('trace_id', generate_span_id())
        results = []

        for span in spans:
            name = span.get('name', 'unknown')
            value = span.get('value', 0)
            rating = span.get('rating', 'unknown')
            page = span.get('page', 'unknown')
            timestamp = span.get('timestamp', 0)
            parent_span_id = span.get('parent_span_id')

            span_id = generate_span_id()
            trace_type = get_trace_type(name)
            start_time = timestamp / 1000
            end_time = start_time + (value / 1000)

            trace = PerformanceTrace.create(
                trace_id=trace_id,
                parent_span_id=parent_span_id,
                span_id=span_id,
                trace_type=trace_type,
                name=name,
                page=page,
                duration=float(value),
                start_time=start_time,
                end_time=end_time,
                rating=rating,
                metadata=json.dumps(span.get('metadata', {}))
            )

            results.append({
                'name': name,
                'span_id': span_id,
                'duration': value
            })

        logger.info(f"[PERF] Batch recorded {len(results)} spans for trace {trace_id[:8]}")

        return jsonify({
            'success': True,
            'trace_id': trace_id,
            'results': results
        }), 200

    except Exception as e:
        logger.error(f"[PERF] Error batch recording: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@performance_bp.route('/trace/<trace_id>', methods=['GET'])
def get_trace_chain(trace_id):
    """获取完整的链路追踪"""
    try:
        # 获取该trace_id的所有span
        traces = PerformanceTrace.select().where(
            PerformanceTrace.trace_id == trace_id
        ).order_by(PerformanceTrace.start_time)

        if not traces:
            return jsonify({'success': False, 'error': 'Trace not found'}), 404

        # 构建树形结构
        span_map = {}
        root_spans = []

        for trace in traces:
            span = {
                'span_id': trace.span_id,
                'parent_span_id': trace.parent_span_id,
                'name': trace.name,
                'trace_type': trace.trace_type,
                'duration': trace.duration,
                'start_time': trace.start_time,
                'end_time': trace.end_time,
                'rating': trace.rating,
                'metadata': json.loads(trace.metadata or '{}'),
                'children': []
            }
            span_map[trace.span_id] = span

        # 构建父子关系
        for span in span_map.values():
            parent_id = span['parent_span_id']
            if parent_id and parent_id in span_map:
                span_map[parent_id]['children'].append(span)
            else:
                root_spans.append(span)

        # 计算总耗时
        total_duration = max(s['end_time'] for s in span_map.values()) - min(s['start_time'] for s in span_map.values())

        return jsonify({
            'success': True,
            'trace_id': trace_id,
            'total_duration': total_duration * 1000,
            'span_count': len(span_map),
            'root_spans': root_spans,
            'created_at': min(t.created_at.timestamp() for t in traces) if traces else None
        })

    except Exception as e:
        logger.error(f"[PERF] Error getting trace: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@performance_bp.route('/recent', methods=['GET'])
def get_recent_traces():
    """获取最近的链路追踪列表"""
    try:
        limit = min(int(request.args.get('limit', 20)), 100)
        offset = int(request.args.get('offset', 0))  # 翻页偏移量
        sort_by = request.args.get('sort_by', 'created_at')  # 支持: created_at, duration
        sort_order = request.args.get('sort_order', 'desc')  # 支持: asc, desc
        operation = request.args.get('operation')  # 操作名称过滤（模糊匹配链路中任意span）
        page = request.args.get('page')  # 页面过滤
        trace_id = request.args.get('trace_id')  # 按 trace_id 搜索（支持完整ID或前缀匹配）

        # 获取所有记录用于分组
        query = PerformanceTrace.select().order_by(PerformanceTrace.created_at.desc()).limit(limit * 20)

        # 按 trace_id 分组，构建链路信息
        trace_groups = {}
        for trace in query:
            tid = trace.trace_id
            if tid not in trace_groups:
                trace_groups[tid] = {
                    'trace_id': tid,
                    'page': trace.page,
                    'start_time': trace.start_time,
                    'created_at': trace.created_at.timestamp() if trace.created_at else None,
                    'duration': trace.duration,
                    'rating': trace.rating,
                    'span_count': 0,
                    'root_name': trace.name,
                    'operations': set(),  # 收集链路中所有操作名
                    'total_duration': 0,  # 链路总耗时
                    'max_duration': 0  # 单个span最大耗时
                }
            group = trace_groups[tid]
            group['span_count'] += 1
            group['operations'].add(trace.name)
            group['total_duration'] += trace.duration
            if trace.duration > group['max_duration']:
                group['max_duration'] = trace.duration
                group['duration'] = trace.duration
                group['rating'] = trace.rating

        # 转换为列表并应用过滤
        result = []
        for group in trace_groups.values():
            group['operations'] = list(group['operations'])  # 转换为列表

            # 按 trace_id 搜索（支持完整ID或前缀匹配）
            if trace_id:
                group_tid = group['trace_id']
                # 支持精确匹配或前缀匹配（如用户输入前8位也能搜索到）
                if trace_id != group_tid and not group_tid.startswith(trace_id):
                    continue

            # 按页面过滤（链路的任一页面匹配即可）
            if page and page not in group.get('page', ''):
                continue

            # 按操作名称模糊过滤（链路的任一操作匹配即可）
            if operation:
                matched = any(operation.lower() in op.lower() for op in group['operations'])
                if not matched:
                    continue

            result.append(group)

        # 按指定字段排序
        reverse = sort_order == 'desc'
        if sort_by == 'duration':
            result.sort(key=lambda x: x['duration'], reverse=reverse)
        else:
            result.sort(key=lambda x: x['start_time'], reverse=reverse)

        # 应用分页
        total = len(result)
        paginated_result = result[offset:offset + limit]

        return jsonify({
            'success': True,
            'traces': paginated_result,
            'total': total,
            'offset': offset,
            'limit': limit,
            'filters': {
                'operation': operation,
                'page': page,
                'trace_id': trace_id,
                'sort_by': sort_by,
                'sort_order': sort_order
            }
        })

    except Exception as e:
        logger.error(f"[PERF] Error getting recent traces: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@performance_bp.route('/stats', methods=['GET'])
def get_performance_stats():
    """获取性能统计"""
    try:
        hours = int(request.args.get('hours', 24))
        name = request.args.get('name')

        # 计算时间范围
        since = datetime.now() - timedelta(hours=hours)

        query = PerformanceTrace.select().where(
            PerformanceTrace.created_at >= since
        )

        if name:
            query = query.where(PerformanceTrace.name == name)

        traces = list(query)

        if not traces:
            return jsonify({
                'success': True,
                'stats': {
                    'count': 0,
                    'avg_duration': 0,
                    'p50': 0,
                    'p90': 0,
                    'p99': 0
                }
            })

        # 计算统计
        durations = sorted([t.duration for t in traces])
        count = len(durations)

        def percentile(data, q):
            idx = int(len(data) * q / 100)
            return data[idx] if idx < len(data) else data[-1]

        stats = {
            'count': count,
            'avg_duration': sum(durations) / count,
            'min_duration': min(durations),
            'max_duration': max(durations),
            'p50': percentile(durations, 50),
            'p90': percentile(durations, 90),
            'p99': percentile(durations, 99),
            'good_count': sum(1 for t in traces if t.rating == 'good'),
            'needs_improvement_count': sum(1 for t in traces if t.rating == 'needs-improvement'),
            'poor_count': sum(1 for t in traces if t.rating == 'poor'),
            'by_name': {}
        }

        # 按名称分组统计
        name_groups = {}
        for trace in traces:
            if trace.name not in name_groups:
                name_groups[trace.name] = []
            name_groups[trace.name].append(trace.duration)

        for name, durs in name_groups.items():
            durs = sorted(durs)
            stats['by_name'][name] = {
                'count': len(durs),
                'avg': sum(durs) / len(durs),
                'p50': percentile(durs, 50),
                'p90': percentile(durs, 90),
                'p99': percentile(durs, 99)
            }

        return jsonify({
            'success': True,
            'stats': stats,
            'time_range': f'{hours} hours'
        })

    except Exception as e:
        logger.error(f"[PERF] Error getting stats: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@performance_bp.route('/clear', methods=['POST'])
def clear_performance_data():
    """清理性能数据"""
    try:
        clear_all = request.args.get('clear_all', 'false').lower() == 'true'
        days = int(request.args.get('days', 7))

        if clear_all:
            # 清空所有链路追踪数据
            deleted = PerformanceTrace.delete().execute()
            logger.info(f"[PERF] Cleared all {deleted} performance records")
        else:
            # 只删除指定天数前的数据
            cutoff = datetime.now() - timedelta(days=days)
            deleted = PerformanceTrace.delete().where(
                PerformanceTrace.created_at < cutoff
            ).execute()
            logger.info(f"[PERF] Cleared {deleted} performance records older than {days} days")

        return jsonify({
            'success': True,
            'deleted': deleted,
            'clear_all': clear_all
        })

    except Exception as e:
        logger.error(f"[PERF] Error clearing data: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

File: src/services/performance.js
/**
 * Performance Trace Service for Electron
 * 
 * Features:
 * - Distributed tracing with span tree structure
 * - Performance metrics collection
 * - Trace chain tracking
 * - Performance statistics
 */

import { settingsDB } from './database.js';

// Trace span status
export const SpanStatus = {
  OK: 'ok',
  ERROR: 'error',
  PENDING: 'pending',
};

// Performance rating thresholds (ms)
const RATING_THRESHOLDS = {
  good: 1000,      // < 1s: good
  needsImprovement: 2500,  // < 2.5s: needs improvement
  // >= 2.5s: poor
};

/**
 * Get performance rating based on duration
 */
function getRating(duration) {
  if (duration < RATING_THRESHOLDS.good) return 'good';
  if (duration < RATING_THRESHOLDS.needsImprovement) return 'needs-improvement';
  return 'poor';
}

/**
 * Generate unique ID
 */
function generateId() {
  return Math.random().toString(36).substring(2, 15) + 
         Math.random().toString(36).substring(2, 15);
}

/**
 * Performance Span
 */
export class PerformanceSpan {
  constructor(name, options = {}) {
    this.spanId = generateId();
    this.traceId = options.traceId || generateId();
    this.parentSpanId = options.parentSpanId || null;
    this.name = name;
    this.startTime = Date.now();
    this.endTime = null;
    this.duration = 0;
    this.status = SpanStatus.PENDING;
    this.metadata = options.metadata || {};
    this.children = [];
  }

  finish(status = SpanStatus.OK, metadata = {}) {
    this.endTime = Date.now();
    this.duration = this.endTime - this.startTime;
    this.status = status;
    Object.assign(this.metadata, metadata);
    return this;
  }

  addChild(childSpan) {
    this.children.push(childSpan);
  }

  toJSON() {
    return {
      span_id: this.spanId,
      trace_id: this.traceId,
      parent_span_id: this.parentSpanId,
      name: this.name,
      start_time: this.startTime,
      end_time: this.endTime,
      duration: this.duration,
      status: this.status,
      rating: getRating(this.duration),
      metadata: this.metadata,
      children: this.children.map(c => c.toJSON()),
    };
  }
}

/**
 * Performance Trace
 */
export class PerformanceTrace {
  constructor(rootName, options = {}) {
    this.traceId = generateId();
    this.rootName = rootName;
    this.page = options.page || '';
    this.createdAt = Math.floor(Date.now() / 1000);
    this.spans = new Map();
    this.rootSpan = null;
  }

  startSpan(name, options = {}) {
    const span = new PerformanceSpan(name, {
      traceId: this.traceId,
      parentSpanId: options.parentSpanId,
      metadata: options.metadata,
    });
    
    this.spans.set(span.spanId, span);
    
    if (!this.rootSpan) {
      this.rootSpan = span;
    }
    
    return span;
  }

  finish() {
    if (this.rootSpan) {
      this.rootSpan.finish();
    }
  }

  toJSON() {
    const allSpans = Array.from(this.spans.values());
    const totalDuration = this.rootSpan?.duration || 0;
    
    // Build span tree
    const spanMap = new Map();
    allSpans.forEach(span => {
      spanMap.set(span.spanId, {
        ...span.toJSON(),
        children: [],
      });
    });
    
    // Build tree structure
    const rootSpans = [];
    spanMap.forEach((span, spanId) => {
      if (span.parent_span_id && spanMap.has(span.parent_span_id)) {
        spanMap.get(span.parent_span_id).children.push(span);
      } else {
        rootSpans.push(span);
      }
    });

    // Get all operation names
    const operations = allSpans.map(s => s.name);

    return {
      trace_id: this.traceId,
      root_name: this.rootName,
      page: this.page,
      created_at: this.createdAt,
      total_duration: totalDuration,
      span_count: allSpans.length,
      rating: getRating(totalDuration),
      operations,
      root_spans: rootSpans,
    };
  }
}

/**
 * Performance Service
 */
export class PerformanceService {
  constructor() {
    this._traces = [];
    this._activeTraces = new Map();
    this._maxTraces = 1000; // Keep last 1000 traces
  }

  /**
   * Initialize service
   */
  async init() {
    await this._loadTraces();
  }

  /**
   * Load traces from database
   */
  async _loadTraces() {
    try {
      const saved = await settingsDB.get('performanceTraces');
      if (saved && Array.isArray(saved)) {
        this._traces = saved;
      }
    } catch (error) {
      console.error('[Performance] Error loading traces:', error);
      this._traces = [];
    }
  }

  /**
   * Save traces to database
   */
  async _saveTraces() {
    try {
      // Trim to max traces
      if (this._traces.length > this._maxTraces) {
        this._traces = this._traces.slice(-this._maxTraces);
      }
      await settingsDB.set('performanceTraces', this._traces);
    } catch (error) {
      console.error('[Performance] Error saving traces:', error);
    }
  }

  /**
   * Start a new trace
   */
  startTrace(rootName, options = {}) {
    const trace = new PerformanceTrace(rootName, options);
    this._activeTraces.set(trace.traceId, trace);
    return trace;
  }

  /**
   * Finish and save a trace
   */
  async finishTrace(traceId) {
    const trace = this._activeTraces.get(traceId);
    if (!trace) return null;

    trace.finish();
    const traceData = trace.toJSON();
    
    this._traces.push(traceData);
    this._activeTraces.delete(traceId);
    
    await this._saveTraces();
    return traceData;
  }

  /**
   * Record performance data directly
   */
  async recordPerformance(data) {
    const traceData = {
      trace_id: data.trace_id || generateId(),
      root_name: data.name || 'unknown',
      page: data.page || '',
      created_at: Math.floor(Date.now() / 1000),
      total_duration: data.duration || 0,
      span_count: 1,
      rating: getRating(data.duration || 0),
      operations: [data.name || 'unknown'],
      root_spans: [{
        span_id: generateId(),
        trace_id: data.trace_id || generateId(),
        parent_span_id: null,
        name: data.name || 'unknown',
        start_time: Date.now() - (data.duration || 0),
        end_time: Date.now(),
        duration: data.duration || 0,
        status: data.status || SpanStatus.OK,
        rating: getRating(data.duration || 0),
        metadata: data.metadata || {},
        children: [],
      }],
    };

    this._traces.push(traceData);
    await this._saveTraces();
    return traceData;
  }

  /**
   * Batch record performance data
   */
  async batchRecordPerformance(items) {
    const results = [];
    for (const item of items) {
      const result = await this.recordPerformance(item);
      results.push(result);
    }
    return results;
  }

  /**
   * Get trace by ID
   */
  getTrace(traceId) {
    return this._traces.find(t => t.trace_id === traceId) || null;
  }

  /**
   * Get recent traces with filtering
   */
  getRecentTraces(options = {}) {
    let traces = [...this._traces];

    // Filter by page
    if (options.page) {
      traces = traces.filter(t => t.page === options.page);
    }

    // Filter by operation
    if (options.operation) {
      const op = options.operation.toLowerCase();
      traces = traces.filter(t => 
        t.operations.some(o => o.toLowerCase().includes(op))
      );
    }

    // Filter by trace_id
    if (options.trace_id) {
      traces = traces.filter(t => 
        t.trace_id.toLowerCase().includes(options.trace_id.toLowerCase())
      );
    }

    // Sort
    const sortBy = options.sort_by || 'created_at';
    const sortOrder = options.sort_order || 'desc';
    
    traces.sort((a, b) => {
      let aVal, bVal;
      if (sortBy === 'duration') {
        aVal = a.total_duration;
        bVal = b.total_duration;
      } else {
        aVal = a.created_at;
        bVal = b.created_at;
      }
      return sortOrder === 'asc' ? aVal - bVal : bVal - aVal;
    });

    // Pagination
    const limit = options.limit || 50;
    const offset = options.offset || 0;
    const total = traces.length;
    traces = traces.slice(offset, offset + limit);

    return { traces, total };
  }

  /**
   * Get performance statistics
   */
  getStats() {
    const traces = this._traces;
    const count = traces.length;

    if (count === 0) {
      return {
        count: 0,
        avg_duration: 0,
        p50: 0,
        p90: 0,
        p99: 0,
        good_count: 0,
        needs_improvement_count: 0,
        poor_count: 0,
        by_name: {},
      };
    }

    const durations = traces.map(t => t.total_duration).sort((a, b) => a - b);
    
    // Calculate percentiles
    const p50 = this._percentile(durations, 0.5);
    const p90 = this._percentile(durations, 0.9);
    const p99 = this._percentile(durations, 0.99);
    
    // Calculate average
    const avg_duration = durations.reduce((a, b) => a + b, 0) / count;
    
    // Count by rating
    const good_count = traces.filter(t => t.rating === 'good').length;
    const needs_improvement_count = traces.filter(t => t.rating === 'needs-improvement').length;
    const poor_count = traces.filter(t => t.rating === 'poor').length;

    // Stats by operation name
    const byName = {};
    traces.forEach(t => {
      const name = t.root_name;
      if (!byName[name]) {
        byName[name] = { count: 0, total_duration: 0, durations: [] };
      }
      byName[name].count++;
      byName[name].total_duration += t.total_duration;
      byName[name].durations.push(t.total_duration);
    });

    // Calculate stats for each operation
    const by_name = {};
    Object.keys(byName).forEach(name => {
      const data = byName[name];
      const sortedDurations = data.durations.sort((a, b) => a - b);
      by_name[name] = {
        count: data.count,
        avg: data.total_duration / data.count,
        p99: this._percentile(sortedDurations, 0.99),
      };
    });

    return {
      count,
      avg_duration,
      p50,
      p90,
      p99,
      good_count,
      needs_improvement_count,
      poor_count,
      by_name,
    };
  }

  /**
   * Calculate percentile
   */
  _percentile(sortedValues, percentile) {
    if (sortedValues.length === 0) return 0;
    const index = Math.ceil(sortedValues.length * percentile) - 1;
    return sortedValues[Math.max(0, index)];
  }

  /**
   * Clear all performance data
   */
  async clearAllData() {
    this._traces = [];
    this._activeTraces.clear();
    await this._saveTraces();
  }

  /**
   * Get trace chain (for detailed view)
   */
  getTraceChain(traceId) {
    const trace = this.getTrace(traceId);
    if (!trace) return null;

    // Flatten span tree for display
    const spans = [];
    const flatten = (spanList, depth = 0) => {
      for (const span of spanList) {
        spans.push({ ...span, depth });
        if (span.children?.length) {
          flatten(span.children, depth + 1);
        }
      }
    };
    flatten(trace.root_spans);

    return {
      ...trace,
      spans,
    };
  }
}

// Singleton instance
let performanceService = null;

export function getPerformanceService() {
  if (!performanceService) {
    performanceService = new PerformanceService();
  }
  return performanceService;
}

export function resetPerformanceService() {
  if (performanceService) {
    performanceService.clearAllData();
    performanceService = null;
  }
}

export default {
  PerformanceService,
  PerformanceTrace,
  PerformanceSpan,
  SpanStatus,
  getPerformanceService,
  resetPerformanceService,
};
